docs/design/2020-06-24-placement-rules-in-sql.md

```diff
@@ -20,401 +51,399 @@ The scenarios of defining placement rules in SQL include:
+Typically, key format is in such a format: `t_{table_id}_r_{pk_value}`, where `pk_value` may be `_tidb_rowid` in some cases. `table_id` can be inferred from `table_name`, thus the key range is `t_{table_id}_` to `t_{table_id+1}_`.
 
-### Database placement
+Similarly, the key range of partitions can also be inferred.
 
-Defining placement rules of databases simplifies the procedures when there are many tables.
+#### Policy Validation
 
-For example, in a typical multi-tenant scenario, each user has a private database. The dataset in one database is relatively small, and it’s rare to query across databases. In this case, a whole database can be placed in a single region to reduce multi-region latency.
+When placement policies are specified, they should be validated for correctness:
 
-For another example, multiple businesses may run on a single TiDB cluster, which can reduce the overhead of maintaining multiple clusters. The resources of multiple businesses need to be isolated to avoid the risk that one business takes too many resources and affects others.
+1. The `FOLLOWERS` count should respect raft quorum expectations. The default is `2` (which creates raft groups of 3). If the number is odd, it could lead to split brain scenarios, so a warning should be issued. Warnings should also be issued for a count less than 2 (this might be useful for development environments, so an error is not returned)
+2. A policy that is impossible based on the current topology (region=us-east-1 and followers=2, but there is only 1 store in us-east-1) should be a warning. This allows for some transitional topologies.
+3. If the constraints are specified as a dictionary, specifying the count (i.e. `FOLLOWERS=n`) is prohibited.
+4. Specifying both direct placement rules (`{FOLLOWERS,VOTERS,LEARNERS}=N, {FOLLOWER,VOTER,LEARNER}_CONSTRAINTS, CONSTRAINTS, PRIMARY_REGION, REGIONS, SCHEDULE`) and a `PLACEMENT POLICY` is prohibited.
+
+#### Skipping Policy Validation
 
-Placement of databases is defined through `ALTER` statements:
+It should be possible to skip policy validation. This can be seen as similar to skipping foreign key checks, which is often used by logical dumpers:
 
 ```sql
-ALTER {DATABASE | SCHEMA} schema_name
-	{ADD | ALTER} PLACEMENT POLICY ROLE=role CONSTRAINTS=constraints REPLICAS=replicas,
-	...
-	
-ALTER {DATABASE | SCHEMA} schema_name
-	DROP PLACEMENT POLICY ROLE=role,
-	...
+SET FOREIGN_KEY_CHECKS=0;
+SET PLACEMENT_CHECKS=0;
+
+CREATE TABLE t3 (a int) PLACEMENT POLICY `mycompanypolicy`;
 ```
 
-This statement defines placement rules for one database, including all tables in it.
+If a table is imported when `PLACEMENT_CHECKS` is `OFF`, and the placement policy does not validate, then the same rules of fallback apply as in the case `DROP PLACEMENT POLICY` (where policy is still in use).
```

Comment:
if we do

```sql
set @@placement_checks = off;
create table t5 (a int) placement policy p6;
-- ...
create placement policy p6 ...;
```

is the placement policy immediately applied to table `t5` once the `create placement policy p6` statement is executed?

Reply:
No, I would suggest the information is lost. There is a similar case where `DROP PLACEMENT POLICY p6` resets all the tables that have this policy to the default.

This will mean that logical importers should restore placement policies first. This makes sense logically anyway, because otherwise you would need to reorganize the data after importing it.